llvm-external-projects/iree-dialects/python/mlir/dialects/iree_pydm/rtl/modules/numerics.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""RTL functions for numeric operations (on scalars)."""
+
+from .constants import *
+from .macros import *
+from ..base import RtlModule
+
+RTL_MODULE = RtlModule("numerics")
+
+
+@RTL_MODULE.export_pyfunc
+def dynamic_binary_promote(left, right) -> tuple:
+  left_order = get_numeric_promotion_order(left)
+  right_order = get_numeric_promotion_order(right)
+  # Note that since we are defining the numeric promotion rules, we have to
+  # use raw functions to compare (or else we would be using the thing we are
+  # defining).
+  if raw_compare_eq(left_order, right_order):
+    return left, right
+  elif raw_compare_gt(left_order, right_order):
+    return left, _promote_to(get_type_code(left), right)
+  else:
+    return _promote_to(get_type_code(right), left), right
+
+
+@RTL_MODULE.internal_pyfunc
```

Comment:
this is really slick - shows how the rtl can grow to be arbitrarily complex without exposing it all or hand-inlining everything thanks to relying on the compiler to inline/dce/etc just as with the user python coming in.

Reply:
+1